src/macaron/slsa_analyzer/ci_service/github_actions/github_actions_ci.py

```diff
@@ -243,14 +244,69 @@ def has_latest_run_passed(
 
         return ""
 
+    def check_publish_start_commit_timestamps(
+        self, started_at: datetime, publish_date_time: datetime, commit_date_time: datetime, time_range: int
+    ) -> bool:
+        """
+        Check if the timestamps of CI run, artifact publishing, and commit date are within the acceptable time range and valid.
+
+        This function checks that the CI run has happened before the artifact publishing timestamp.
+
+        This function also verifies whether the commit date is within an acceptable time range
+        from the publish start time. The acceptable range is defined as half of the provided
+        time range parameter.
+
+        Parameters
+        ----------
+        started_at : datetime
+            The timestamp indicating when the GitHub Actions workflow started.
+        publish_date_time : datetime
+            The timestamp indicating when the artifact is published.
+        commit_date_time : datetime
+            The timestamp of the source code commit.
+        time_range : int
+            The total acceptable time range in seconds.
+
+        Returns
+        -------
+        bool
+            True if the commit date is within the acceptable range from the publish start time,
+                False otherwise. Returns False in case of any errors during timestamp comparisons.
+        """
+        try:
+            if started_at < publish_date_time:
+                # Make sure the source-code commit date is also within acceptable range.
+                acceptable_range = time_range / 2
+                if timedelta.total_seconds(abs(started_at - commit_date_time)) > acceptable_range:
+                    logger.debug(
+                        (
+                            "The difference between GitHub Actions starting time %s and source commit time %s"
+                            " is not within %s seconds."
+                        ),
+                        started_at,
+                        commit_date_time,
+                        acceptable_range,
+                    )
+                    return False
+                return True
+
+        # Handle errors for calls to `fromisoformat()` and the time comparison.
```

Comment:
I don't think `fromisoformat` is used here?

Reply:
Yes, this was left from another change:. See commit [c9761dc](https://github.com/oracle/macaron/pull/872/commits/c9761dca9c853e65a196c033ebe45b53bbf3ac05).